cmd/transcriber/call/tracks.go

```diff
@@ -0,0 +1,341 @@
+}
+
+// decodeAudio reads a track OGG file and decodes its audio into raw PCM samples
+// for later processing.
+func (ctx trackContext) decodeAudio() ([]trackTimedSamples, error) {
+	trackFile, err := os.Open(ctx.filename)
+	defer trackFile.Close()
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to open track file: %w", err)
+	}
+
+	oggReader, _, err := ogg.NewReaderWith(trackFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new ogg reader: %w", err)
+	}
+
+	opusDec, err := opus.NewDecoder(trackOutAudioRate, trackAudioChannels)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create opus decoder: %w", err)
+	}
+	defer func() {
+		if err := opusDec.Destroy(); err != nil {
+			log.Printf("failed to destroy decoder: %s", err)
+		}
+	}()
+
+	log.Printf("decoding track %s", ctx.trackID)
+
+	pcmBuf := make([]float32, trackOutFrameSize)
+	// TODO: consider pre-calculating track duration to minimize memory waste.
+	samples := make([]trackTimedSamples, 1)
+
+	var prevGP uint64
+	for {
+		data, hdr, err := oggReader.ParseNextPage()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			log.Printf("failed to parse off page: %s", err)
+			continue
+		}
+
+		// Ignoring first page which only contains metadata.
+		if hdr.GranulePosition == 0 {
+			continue
+		}
+
+		if hdr.GranulePosition > prevGP+uint64(trackInFrameSize) {
+			log.Printf("%v gap in audio samples", time.Duration((hdr.GranulePosition-prevGP)/trackInAudioSamplesPerMs)*time.Millisecond)
+			samples = append(samples, trackTimedSamples{
+				startTS: int64(hdr.GranulePosition) / trackInAudioSamplesPerMs,
+			})
+		}
+		prevGP = hdr.GranulePosition
+
+		n, err := opusDec.Decode(data, pcmBuf)
+		if err != nil {
+			log.Printf("failed to decode audio data: %s", err)
+		}
+
+		samples[len(samples)-1].pcm = append(samples[len(samples)-1].pcm, pcmBuf[:n]...)
+	}
+
+	return samples, nil
+}
+
+// transcribeTrack feeds track's raw audio samples to a transcription engine (e.g. whisper)
+// and outputs a transcription.
+func (t *Transcriber) transcribeTrack(ctx trackContext) (transcribe.TrackTranscription, time.Duration, error) {
+	trackTr := transcribe.TrackTranscription{
+		Speaker: ctx.user.GetDisplayName(model.ShowFullName),
+	}
+
+	samples, err := ctx.decodeAudio()
+	if err != nil {
+		return trackTr, 0, fmt.Errorf("failed to decode audio samples: %w", err)
+	}
+
+	transcriber, err := t.newTrackTranscriber()
+	if err != nil {
+		return trackTr, 0, fmt.Errorf("failed to create track transcriber: %w", err)
+	}
+
+	var totalDur time.Duration
+	for _, ts := range samples {
+		segments, err := transcriber.Transcribe(ts.pcm)
+		if err != nil {
+			log.Printf("failed to transcribe audio samples: %s", err)
+			continue
+		}
+
+		samplesDur := time.Duration(len(ts.pcm)/trackOutAudioSamplesPerMs) * time.Millisecond
+		totalDur += samplesDur
+
+		for _, s := range segments {
+			s.StartTS += ts.startTS + ctx.startTS
+			s.EndTS += ts.startTS + ctx.startTS
```

Comment:
This is a little tricky to keep track of (the relative offsets and incrementing them here like this). Is there a clearer way to do it? If not, no worries, just thought I'd point out where I had to think about things for a bit.

Reply:
I think it's just inherently tricky. No way to simplify I can see other than naming them differently as we are dealing with at least 3 independent timestamps. I'll look into this though, but will need some time.